src/snowflake/snowpark/_internal/data_source/datasource_partitioner.py

```diff
@@ -0,0 +1,246 @@
+#
+# Copyright (c) 2012-2025 Snowflake Computing Inc. All rights reserved.
+#
+
+import datetime
+import decimal
+from _decimal import ROUND_HALF_EVEN, ROUND_HALF_UP
+from typing import Optional, Union, List, Callable
+
+import pytz
+from dateutil import parser
+from snowflake.snowpark._internal.data_source.utils import (
+    detect_dbms,
+    DBMS_MAP,
+    DRIVER_MAP,
+)
+
+from snowflake.snowpark._internal.data_source.datasource_reader import DataSourceReader
+import logging
+
+from snowflake.snowpark._internal.type_utils import type_string_to_type_object
+from snowflake.snowpark._internal.data_source.datasource_typing import Connection
+from snowflake.snowpark.types import (
+    StructType,
+    _NumericType,
+    DateType,
+    DataType,
+    TimestampType,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class DataSourcePartitioner:
+    def __init__(
+        self,
+        create_connection: Callable[[], "Connection"],
+        table_or_query: str,
+        column: Optional[str] = None,
+        lower_bound: Optional[Union[str, int]] = None,
+        upper_bound: Optional[Union[str, int]] = None,
+        num_partitions: Optional[int] = None,
+        query_timeout: Optional[int] = 0,
+        fetch_size: Optional[int] = 0,
+        custom_schema: Optional[Union[str, StructType]] = None,
+        predicates: Optional[List[str]] = None,
+    ) -> None:
+        self.create_connection = create_connection
+        self.table_or_query = table_or_query
+        self.column = column
+        self.lower_bound = lower_bound
+        self.upper_bound = upper_bound
+        self.num_partitions = num_partitions
+        self.query_timeout = query_timeout
+        self.fetch_size = fetch_size
+        self.custom_schema = custom_schema
+        self.predicates = predicates
+        conn = create_connection()
+        dbms_type, driver = detect_dbms(conn)
+        self.dialect = DBMS_MAP[dbms_type]()
+        self.driver = DRIVER_MAP[driver](create_connection)
+        self.dialect_class = DBMS_MAP[dbms_type]
+        self.driver_class = DRIVER_MAP[driver]
```

Comment:
nit: you can do the following to make it slightly clear
```suggestion
        self.dialect_class = DBMS_MAP[dbms_type]
        self.driver_class = DRIVER_MAP[driver]
        self.dialect = dialect_class()
        self.driver = driver_class(create_connection)
```